### PR TITLE
Handles a `PushRepository.DoesNotExist` exception

### PR DESCRIPTION
**Pull request by git-hyagi**
fixes: #1712

### CODE REVIEW

File: pulp_container/app/registry_api.py [hunk 1]
```diff
@@ -392,7 +392,16 @@ def get_dr_push(self, request, path, create=False):
...
-            repository = serializers.ContainerPushRepositorySerializer.get_or_create({"name": path})
+            try:
+                repository = serializers.ContainerPushRepositorySerializer.get_or_create(
+                    {"name": path}
+                )
+            except models.ContainerPushRepository.DoesNotExist:
+                raise RepositoryInvalid(
+                    name=path,
+                    message='Cannot create "' + path + '" push repository, another repository with '
+                    "the same name already exists.",
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:403]: Too much information. Let's stick to raising just `RepositoryInvalid(name=path, message="Repository is read-only.")`.

File: docs/user/guides/push-image.md [hunk 2]
```diff
@@ -47,7 +47,8 @@ X-Frame-Options: SAMEORIGIN
 !!! note
 
     Content is pushed to a push repository type. A push repository does not support mirroring of the
-    remote content via the Pulp API.
+    remote content via the Pulp API. Trying to push a content with the same name of an existing
+    "regular" repository will fail.
```
Comment (lubosmj) [inline docs/user/guides/push-image.md:51]:
```suggestion
    remote content via the Pulp API. Trying to push content with the same name as an existing
    "regular" repository will fail.
```

File: CHANGES/1712.bugfix [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Resolved an issue that caused an HTTP 500 error when attempting to push an image with the same name
+as an existing regular repository.
```
Comment (lubosmj) [inline CHANGES/1712.bugfix:2]:
```suggestion
Fixed an HTTP 500 error returned when pushing an image with the same name as an existing regular repository.
```

File: pulp_container/app/registry_api.py [hunk 4]
```diff
@@ -392,7 +392,16 @@ def get_dr_push(self, request, path, create=False):
...
+                )
             distribution = serializers.ContainerDistributionSerializer.get_or_create(
                 {"base_path": path, "name": path}, {"repository": get_url(repository)}
             )
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:407]: We also need to catch an exception here.
```
pulp file distribution create --name test/mygroup/ubuntu --base-path test/mygroup/ubuntu
podman push localhost:5001/test/mygroup/ubuntu --tls-verify=false
```
```
[pulp]           | During handling of the above exception, another exception occurred:
[pulp]           | 
[pulp]           | Traceback (most recent call last):
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
[pulp]           |     response = get_response(request)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
[pulp]           |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
[pulp]           |     return view_func(*args, **kwargs)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 124, in view
[pulp]           |     return self.dispatch(request, *args, **kwargs)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
[pulp]           |     response = self.handle_exception(exc)
[pulp]           |   File "/src/pulp_container/pulp_container/app/registry_api.py", line 279, in handle_exception
[pulp]           |     response = super().handle_exception(exc)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
[pulp]           |     self.raise_uncaught_exception(exc)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
[pulp]           |     raise exc
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
[pulp]           |     response = handler(request, *args, **kwargs)
[pulp]           |   File "/src/pulp_container/pulp_container/app/registry_api.py", line 773, in create
[pulp]           |     _, repository = self.get_dr_push(request, path, create=True)
[pulp]           |   File "/src/pulp_container/pulp_container/app/registry_api.py", line 373, in get_dr_push
[pulp]           |     distribution, repository = self.create_dr(path, request)
[pulp]           |   File "/src/pulp_container/pulp_container/app/registry_api.py", line 405, in create_dr
[pulp]           |     distribution = serializers.ContainerDistributionSerializer.get_or_create(
[pulp]           |   File "/src/pulpcore/pulpcore/app/serializers/base.py", line 381, in get_or_create
[pulp]           |     result = cls.Meta.model.objects.get(**natural_key)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 87, in manager_method
[pulp]           |     return getattr(self.get_queryset(), name)(*args, **kwargs)
[pulp]           |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 637, in get
[pulp]           |     raise self.model.DoesNotExist(
[pulp]           | pulp_container.app.models.ContainerDistribution.DoesNotExist: ContainerDistribution matching query does not exist.
```

Maybe, we should start monitoring both `get_or_create` calls. `django.core.exceptions.ObjectDoesNotExist` would do us a favour.

**Review by lubosmj — CHANGES_REQUESTED** [review]

File: pulp_container/app/registry_api.py [hunk 5]
```diff
@@ -392,10 +392,19 @@ def get_dr_push(self, request, path, create=False):
...
+                )
+            except models.ContainerPushRepository.DoesNotExist:
+                raise RepositoryInvalid(name=path, message="Repository is read-only.")
+
+            try:
+                distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                    {"base_path": path, "name": path}, {"repository": get_url(repository)}
+                )
+            except models.ContainerDistribution.DoesNotExist:
+                raise InvalidRequest(message="ContainerDistribution does not exist.")
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:407]: I guess we can raise again RepositoryInvalid. But, I have no idea what to tell to client...

File: pulp_container/app/registry_api.py [hunk 6]
```diff
@@ -392,10 +392,24 @@ def get_dr_push(self, request, path, create=False):
...
+            try:
+                distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                    {"base_path": path, "name": path}, {"repository": get_url(repository)}
+                )
+            except models.ContainerDistribution.DoesNotExist:
+                raise RepositoryInvalid(
+                    name=path,
+                    message="Cannot push to "
+                    + path
+                    + ". Another plugin with the same distribution name already exists.",
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:411]: What if there is a pulp-container pull-through distribution that has the same name? Is it still another plugin? 😈

Comment (git-hyagi) [reply pulp_container/app/registry_api.py:411]: Good catch!

File: pulp_container/app/registry_api.py [hunk 7]
```diff
@@ -392,10 +392,19 @@ def get_dr_push(self, request, path, create=False):
...
+                )
+            except models.ContainerPushRepository.DoesNotExist:
+                raise RepositoryInvalid(name=path, message="Repository is read-only.")
+
+            try:
+                distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                    {"base_path": path, "name": path}, {"repository": get_url(repository)}
+                )
+            except models.ContainerDistribution.DoesNotExist:
+                raise RepositoryInvalid(name=path, message="Repository is ready-only.")
```
Comment (lubosmj) [inline pulp_container/app/registry_api.py:407]:
```suggestion
            try:
                repository = serializers.ContainerPushRepositorySerializer.get_or_create(
                    {"name": path}
                )
                distribution = serializers.ContainerDistributionSerializer.get_or_create(
                    {"base_path": path, "name": path}, {"repository": get_url(repository)}
                )
            except ObjectDoesNotExist:
                raise RepositoryInvalid(name=path, message="Repository is ready-only.")
```

Comment (git-hyagi) [reply pulp_container/app/registry_api.py:407]: :man_facepalming:

Comment (git-hyagi) [reply pulp_container/app/registry_api.py:407]: hum... actually, it seems like it does not work:
```
...
   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 637, in get
     raise self.model.DoesNotExist(
 pulp_container.app.models.ContainerDistribution.DoesNotExist: ContainerDistribution matching query does not exist.
...
   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 637, in get
     raise self.model.DoesNotExist(
 pulp_container.app.models.ContainerPullThroughRemote.DoesNotExist: ContainerPullThroughRemote matching query does not exist.
```

the django `ObjectDoesNotExist` exception did not catch the `ContainerDistribution.DoesNotExist` and/or `ContainerPullThroughRemote.DoesNotExist` exceptions

Comment (git-hyagi) [reply pulp_container/app/registry_api.py:407]: nvm, the above exception is not related with the proposed change.

We opened a new issue (#1729) to investigate it.

**Review by lubosmj — APPROVED** [review]
Simple and straightforward! 🐁

**Pull request merged**